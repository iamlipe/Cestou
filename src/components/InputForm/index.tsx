import React, {memo, useState, useCallback} from 'react';
import styled, {useTheme} from 'styled-components/native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {TextInputProps} from 'react-native';
import {useController} from 'react-hook-form';
import {
  TextInputMask,
  TextInputMaskTypeProp,
  TextInputMaskOptionProp,
} from 'react-native-masked-text';
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';

interface Props extends TextInputProps {
  name: string;
  control: any;
  label?: string;
  type?: TextInputMaskTypeProp;
  options?: TextInputMaskOptionProp;
  error?: string;
  secureTextEntry?: boolean;
}

const InputForm = ({
  name,
  control,
  label,
  type,
  options,
  error,
  secureTextEntry = false,
}: Props) => {
  const [securityText, setSecurityText] = useState(secureTextEntry);
  const placeholderTraslateX = useSharedValue(0);
  const placeholderFontSize = useSharedValue(14);
  const placeholderOpacity = useSharedValue(1);
  const theme = useTheme();

  const {
    field: {onChange, value},
  } = useController({name, control});

  const handleFocus = useCallback(
    (editing: boolean) => {
      if (editing) {
        placeholderTraslateX.value = withTiming(-12, {duration: 100});
        placeholderFontSize.value = withTiming(12, {duration: 100});
        placeholderOpacity.value = withTiming(0.6, {duration: 100});
      } else {
        placeholderTraslateX.value = withTiming(0, {duration: 100});
        placeholderFontSize.value = withTiming(14, {duration: 100});
        placeholderOpacity.value = withTiming(1, {duration: 100});
      }
    },
    [placeholderFontSize, placeholderOpacity, placeholderTraslateX],
  );

  const animatedStylePlaceholder = useAnimatedStyle(() => {
    return {
      transform: [{translateY: placeholderTraslateX.value}],
      fontSize: placeholderFontSize.value,
      opacity: placeholderOpacity.value,
    };
  });

  return (
    <StyledContainer>
      <StyledContainerInput>
        <StyledLabel style={animatedStylePlaceholder}>{label}</StyledLabel>
        {type ? (
          <StyledInputMask
            testID={`input-text-${name}`}
            type={type}
            options={options}
            onChangeText={onChange}
            onFocus={() => handleFocus(true)}
            onBlur={() => {
              if (value === '') handleFocus(false);
            }}
            value={value}
          />
        ) : (
          <StyledInput
            testID={`input-text-${name}`}
            onChangeText={onChange}
            onFocus={() => handleFocus(true)}
            onBlur={() => {
              if (value === '') handleFocus(false);
            }}
            value={value}
            secureTextEntry={securityText}
          />
        )}

        {secureTextEntry && (
          <Icon.Button
            testID={securityText ? 'visibility-icon' : 'visibility-off-icon'}
            name={securityText ? 'visibility' : 'visibility-off'}
            iconStyle={{marginRight: 0}}
            size={20}
            color={theme.colors.GRAY_900}
            backgroundColor={theme.colors.SUCCESS_100}
            onPress={() => setSecurityText(!securityText)}
          />
        )}
        {error && (
          <Icon
            testID="error-icon"
            name="warning"
            size={24}
            color={theme.colors.ERROR_800}
          />
        )}
      </StyledContainerInput>
      <StyledError>{error}</StyledError>
    </StyledContainer>
  );
};

const StyledContainer = styled.View`
  margin-bottom: 10px;
`;

const StyledContainerInput = styled.View`
  width: 100%;
  height: 56px;
  flex-direction: row;
  align-items: center;
  border-bottom-width: 1px;
  background-color: ${({theme}) => theme.colors.SUCCESS_100};
  border-color: ${({theme}) => theme.colors.GRAY_700};
  padding: 0 16px;
`;

const StyledLabel = styled(Animated.Text)`
  position: absolute;
  padding-left: 16px;
  color: ${({theme}) => theme.colors.GRAY_900};
`;

const StyledError = styled.Text`
  align-self: flex-end;
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLEST};
  color: ${({theme}) => theme.colors.ERROR_800};
  margin-bottom: 3px;
`;

const StyledInput = styled.TextInput`
  flex: 1;
  font-size: ${({theme}) => theme.sizing.SMALLER};
  margin-top: 15px;
`;

const StyledInputMask = styled(TextInputMask)`
  flex: 1;
  font-size: ${({theme}) => theme.sizing.SMALLER};
  margin-top: 15px;
`;

export default memo(InputForm);
