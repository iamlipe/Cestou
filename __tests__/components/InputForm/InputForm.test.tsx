import React from 'react';
import {fireEvent} from '@testing-library/react-native';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/config/functions/testing';

import InputForm from '@/components/InputForm';

describe('InputForm', () => {
  const name = 'default';
  const errorMessage = 'someting went wrong';
  const control = jest.fn();

  test('should render correctly', () => {
    renderWithThemeProvider(<InputForm name={name} control={control} />);
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(<InputForm name={name} control={control} />);
  });

  test('should not render icon visibility when secureTextEntry is false', () => {
    const inputForm = renderWithThemeProvider(
      <InputForm name={name} control={control} />,
    );

    const input = inputForm.getByTestId(`input-text-${name}`);
    const visibilityIcon = inputForm.queryByTestId('visibility-icon');

    expect(visibilityIcon).toBeFalsy();
    expect(input.props.secureTextEntry).toBeFalsy();
  });

  test('should render icon visibility when secureTextEntry is true', () => {
    const inputForm = renderWithThemeProvider(
      <InputForm name={name} control={control} secureTextEntry />,
    );

    const input = inputForm.getByTestId(`input-text-${name}`);
    const visibilityIcon = inputForm.getByTestId('visibility-icon');

    expect(visibilityIcon).toBeTruthy();
    expect(input.props.secureTextEntry).toBeTruthy();
  });

  test('should render icon visibility-off when press button visibilty and return to visibiliy when press again', () => {
    const inputForm = renderWithThemeProvider(
      <InputForm name={name} control={control} secureTextEntry />,
    );

    const input = inputForm.getByTestId(`input-text-${name}`);
    const visibilityIcon = inputForm.getByTestId('visibility-icon');

    expect(visibilityIcon).toBeTruthy();
    expect(input.props.secureTextEntry).toBeTruthy();

    fireEvent.press(visibilityIcon);

    const visibilityOffIcon = inputForm.getByTestId('visibility-off-icon');
    expect(visibilityOffIcon).toBeTruthy();
    expect(input.props.secureTextEntry).toBeFalsy();

    fireEvent.press(visibilityOffIcon);

    expect(visibilityIcon).toBeTruthy();
    expect(input.props.secureTextEntry).toBeTruthy();
  });

  test('should render icon error when pass error to inputForm', () => {
    const inputForm = renderWithThemeProvider(
      <InputForm name={name} control={control} error={errorMessage} />,
    );

    const errorIcon = inputForm.getByTestId('error-icon');

    expect(errorIcon).toBeTruthy();
  });

  test('should render a message with a error when pass a error to inputForm', () => {
    const inputForm = renderWithThemeProvider(
      <InputForm name={name} control={control} error={errorMessage} />,
    );

    const errorText = inputForm.getByText(errorMessage);

    expect(errorText).toBeTruthy();
  });
});
