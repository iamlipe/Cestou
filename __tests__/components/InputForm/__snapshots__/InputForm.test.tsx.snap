// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputForm should matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 10,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f3fbf5",
          "borderBottomWidth": 1,
          "borderColor": "#3A3A3A",
          "flexDirection": "row",
          "height": 56,
          "paddingBottom": 0,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 0,
          "width": "100%",
        },
      ]
    }
  >
    <Text
      animatedStyle={
        Object {
          "value": Object {
            "fontSize": 14,
            "opacity": 1,
            "transform": Array [
              Object {
                "translateY": 0,
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "color": "#1A1A1A",
          "fontSize": 14,
          "opacity": 1,
          "paddingLeft": 16,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
    />
    <TextInput
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onFocus={[Function]}
      secureTextEntry={false}
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 27,
            "marginTop": 15,
          },
        ]
      }
      testID="input-text-default"
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "flex-end",
          "color": "#781B1C",
          "fontFamily": "SourceSansPro-Regular",
          "fontSize": 24,
          "marginBottom": 3,
        },
      ]
    }
  />
</View>
`;
